### PR TITLE
feat!: Allow custom error types to be used.

### PR DESCRIPTION
**Pull request by nzakas**
This adds the ability to override the error classes used when `Env` throws errors.

You can assign a different class to `Env.KeyNotFoundError` and `Env.EmptyStringError` to change the behavior of all `Env` instances. The default errors also have a `name` property that is easy to identify for custom error handling.

@boneskull Can you double-check my typings here? Not sure I got it right.
@FredKSchott This should solve your use case in a couple of different ways, so interested in your feedback.

Closes #77

### CODE REVIEW

**Comment by nzakas** [comment]
Last call for feedback. :)

File: src/env.js [hunk 1]
```diff
@@ -28,24 +28,62 @@ const defaultEnvSource = (() => {
...
- * @throws {Error} Always. 
+ * The error thrown when a required key is not found.
  */
-function keyNotFound(key) {
-    throw new Error(`Required environment variable '${String(key)}' not found.`);
+export class EnvKeyNotFoundError extends Error {
+
+    /**
+     * Creates a new instance.
+     * @param {PropertyKey} key The key that wasn't found.
```
Comment (boneskull) [inline src/env.js:42]:
```suggestion
     * @param {string} key The key that wasn't found. 
```

This should be `string`.  Even though `PropertyKey` is allowed on any object, the keys of `NodeJS.ProcessEnv` are always `string`.

Ref: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/process.d.ts#L112

File: src/env.js [hunk 2]
```diff
@@ -28,24 +28,62 @@ const defaultEnvSource = (() => {
...
-    throw new Error(`Required environment variable '${String(key)}' not found.`);
+export class EnvKeyNotFoundError extends Error {
+
+    /**
+     * Creates a new instance.
+     * @param {PropertyKey} key The key that wasn't found. 
+     */
+    constructor(key) {
+
+        super(`Required environment variable '${String(key)}' not found.`);
```
Comment (boneskull) [inline src/env.js:46]:
```suggestion
        super(`Required environment variable '${key}' not found.`);
```

File: src/env.js [hunk 3]
```diff
@@ -28,24 +28,62 @@ const defaultEnvSource = (() => {
...
+
+        /**
+         * The key that wasn't found.
+         * @type {PropertyKey}
```
Comment (boneskull) [inline src/env.js:50]:
```suggestion
         * @type {string}
```

File: src/env.js [hunk 4]
```diff
@@ -28,24 +28,62 @@ const defaultEnvSource = (() => {
...
- * @throws {Error} Always.
+ * The error thrown when a required key is present but is an empty string.
  */
-function emptyString(key) {
-    throw new Error(`Required environment variable '${String(key)}' is an empty string.`);
+export class EnvEmptyStringError extends Error {
+
+    /**
+     * Creates a new instance.
+     * @param {PropertyKey} key The key that wasn't found.
```
Comment (boneskull) [inline src/env.js:69]:
```suggestion
     * @param {string} key The key that wasn't found. 
```

File: src/env.js [hunk 5]
```diff
@@ -28,24 +28,62 @@ const defaultEnvSource = (() => {
...
-    throw new Error(`Required environment variable '${String(key)}' is an empty string.`);
+export class EnvEmptyStringError extends Error {
+
+    /**
+     * Creates a new instance.
+     * @param {PropertyKey} key The key that wasn't found. 
+     */
+    constructor(key) {
+
+        super(`Required environment variable '${String(key)}' is an empty string.`);
```
Comment (boneskull) [inline src/env.js:73]:
```suggestion
        super(`Required environment variable '${key}' is an empty string.`);
```

File: src/env.js [hunk 6]
```diff
@@ -28,24 +28,62 @@ const defaultEnvSource = (() => {
...
+
+        /**
+         * The key that wasn't found.
+         * @type {PropertyKey}
```
Comment (boneskull) [inline src/env.js:77]:
```suggestion
         * @type {string}
```

**Review by boneskull — CHANGES_REQUESTED** [review]
see note about `PropertyKey` vs `string`

**Review by boneskull — APPROVED** [review]
I guess I probably shouldn't block the PR, since it's your repo. lol

File: src/env.js (hunk 1 above)
Comment (nzakas) [reply src/env.js:42]: Gotcha, makes sense. 👍

**Comment by FredKSchott** [comment]
This is great, thanks for tackling this @nzakas!

**Comment by nzakas** [comment]
Just going to go through and do a few more updates then I'll release this.

**Pull request merged**